### PR TITLE
Enhancement: Use Sandbox for Testing

### PR DESCRIPTION
**Pull request by tzaffi**
## Blocked by https://github.com/algorand/algorand-sdk-testing/pull/220
...which explains most of the removed steps.

## Additional Unused Steps via Unused Steps Analysis Script
according to this [spreadsheet](https://docs.google.com/spreadsheets/d/1Szfvw6_OV0cTz-IEN9bTop5ZW2i472D-pmIHyPVJing/edit?usp=sharing)

* `the transaction should still be the same`
* `I get transactions by address only`
* `I get transactions by address and date`
* `@Given("indexer client {int} at {string} port {int} with token {string}")`
* `@Given("key registration transaction parameters {int} {int} {int} {string} {string} {string} {int} {int} {int} {string} {string}")`
* `@When("I create the payment transaction")`
* `@When("I create the key registration transaction")`
* `@Then("I can get the transaction by ID")`
* `@When("I read a transaction {string} from file {string}")`
* `@When("I write the transaction to file")`
* `@Then("I do my part")`
* `@When("I get recent transactions, limited by {int} transactions")`

### CODE REVIEW

File: test-harness.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  rm -rf $TEST_DIR/unit
+  rm -rf $TEST_RESOURCES_DIR
+fi
+mkdir -p $TEST_DIR/integration
+mkdir -p $TEST_DIR/unit
+mkdir -p $TEST_RESOURCES_DIR
+
+# The Java implementation of these is too tightly coupled with the
+# integration tests, so add them to the integration tests instead.
+mv "$SDK_TESTING_HARNESS"/features/unit/offline.feature test-harness/features/integration/
```
Comment (tzaffi) [inline test-harness.sh:52]:
```suggestion
mv "$SDK_TESTING_HARNESS"/features/unit/offline.feature "$SDK_TESTING_HARNESS/features/integration/
```

File: test-harness.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  rm -rf $TEST_DIR/unit
+  rm -rf $TEST_RESOURCES_DIR
+fi
+mkdir -p $TEST_DIR/integration
+mkdir -p $TEST_DIR/unit
+mkdir -p $TEST_RESOURCES_DIR
+
+# The Java implementation of these is too tightly coupled with the
+# integration tests, so add them to the integration tests instead.
+mv "$SDK_TESTING_HARNESS"/features/unit/offline.feature "$SDK_TESTING_HARNESS/features/integration/
```
Comment (tzaffi) [inline test-harness.sh:52]: @winder - Can you provide a bit of context here? I'm surprised to see a unit test being turned into an integration test.

Comment (winder) [reply test-harness.sh:52]: IIRC while refactoring the step files I couldn't clean it up completely. This one somehow depended on helpers / state used by the integration tests.

File: .test-env [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+VERBOSE_HARNESS=1
+
+# WARNING: If set to 1, new features will be LOST when downloading the test harness.
+# REGARDLESS: modified features are ALWAYS overwritten.
+REMOVE_LOCAL_FEATURES=0
+
+# WARNING: Be careful when turning on the next variable.
+# In that case you'll need to provide all variables expected by `algorand-sdk-testing`'s `.env`
+OVERWRITE_TESTING_ENVIRONMENT=0
```
Comment (tzaffi) [inline .test-env:14]:
```suggestion
OVERWRITE_TESTING_ENVIRONMENT=0

```

File: Makefile [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
 unit:
-	mvn test -Dcucumber.filter.tags="@unit.offline or @unit.algod or @unit.indexer or @unit.rekey or @unit.indexer.rekey or @unit.transactions or @unit.transactions.keyreg or @unit.responses or @unit.applications or @unit.dryrun or @unit.tealsign or @unit.responses.messagepack or @unit.responses.231 or @unit.responses.messagepack.231 or @unit.feetest or @unit.indexer.logs or @unit.abijson or @unit.abijson.byname or @unit.atomic_transaction_composer or @unit.transactions.payment or @unit.responses.unlimited_assets or @unit.algod.ledger_refactoring or @unit.indexer.ledger_refactoring or @unit.dryrun.trace.application or @unit.sourcemap"
+	mvn test -Dcucumber.filter.tags=$(UNITS)
 
+INTEGRATIONS = "@algod or @assets or @auction or @kmd or @send or @send.keyregtxn or @indexer or @rekey_v1 or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @indexer.231 or @abi or @c2c or @compile.sourcemap"
```
Comment (michaeldiamant) [inline Makefile:5]: @tzaffi Based on https://github.com/algorand/go-algorand-sdk/pull/360#discussion_r949337599, I think `@indexer.231` ought to be removed.

Note - I didn't assess which additional steps (if any) can also be removed.

```suggestion
INTEGRATIONS = "@algod or @assets or @auction or @kmd or @send or @send.keyregtxn or @indexer or @rekey_v1 or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @abi or @c2c or @compile.sourcemap"
```

Comment (tzaffi) [reply Makefile:5]: I'm bringing over the same approach as in python and go here, so resolving for now.

File: Makefile [hunk 5]
```diff
@@ -1,10 +1,24 @@
+UNIT_TAGS :=  "$(subst :, or ,$(shell awk '{print $2}' src/test/unit.tags | paste -s -d: -))"
+INTEGRATIONS_TAGS := "$(subst :, or ,$(shell awk '{print $2}' src/test/integration.tags | paste -s -d: -))"
```
Comment (tzaffi) [inline Makefile:2]:
```suggestion
INTEGRATION_TAGS := "$(subst :, or ,$(shell awk '{print $2}' src/test/integration.tags | paste -s -d: -))"
```

Comment (tzaffi) [reply Makefile:2]: dohhhh - thanks @ahangsu

Comment (ahangsu) [reply Makefile]: 🕵️‍♂️ !

File: .test-env [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# Configs for testing repo download:
+SDK_TESTING_URL="https://github.com/algorand/algorand-sdk-testing"
+SDK_TESTING_BRANCH="trim-indexer"
+SDK_TESTING_HARNESS="test-harness"
```
Comment (tzaffi) [inline .test-env:4]: ^^^^ TODO: don't merge as is

File: .test-env [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Configs for testing repo download:
+SDK_TESTING_URL="https://github.com/algorand/algorand-sdk-testing"
+SDK_TESTING_BRANCH="trim-indexer"
```
Comment (tzaffi) [inline .test-env:3]:
```suggestion
SDK_TESTING_BRANCH="master"
```

Should merge this commit before merging the PR.

File: Makefile [hunk 8]
```diff
@@ -1,10 +1,24 @@
...
+	# Build SDK testing environment
+	docker build -t java-sdk-testing .
+
+docker-javasdk-run:
+	# Launch SDK testing
+	docker run -it --network host java-sdk-testing:latest
 
-docker-test:
-	./run_integration_tests.sh
+docker-test: harness docker-javasdk-build docker-javasdk-run
```
Comment (tzaffi) [inline Makefile:24]: Breaking up `make docker-test` into:
1. pull down the `harness`
2. build the javasdk docker
3. run the javasdk docker (which runs `make unit` and `make integration`)

File: Makefile [hunk 9]
```diff
@@ -1,10 +1,24 @@
+UNIT_TAGS :=  "$(subst :, or ,$(shell awk '{print $2}' src/test/unit.tags | paste -s -d: -))"
+INTEGRATION_TAGS := "$(subst :, or ,$(shell awk '{print $2}' src/test/integration.tags | paste -s -d: -))"
```
Comment (tzaffi) [inline Makefile:2]: cucumber tags now reside in:
* `src/test/integration.tags`
* `src/test/unit.tags`

File: Makefile [hunk 10]
```diff
@@ -1,10 +1,24 @@
...
+	mvn test -Dcucumber.filter.tags=$(UNIT_TAGS)
 
 integration:
-	mvn test \
-          -Dtest=com.algorand.algosdk.integration.RunCucumberIntegrationTest \
-          -Dcucumber.filter.tags="@algod or @assets or @auction or @kmd or @send or @send.keyregtxn or @indexer or @rekey_v1 or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @indexer.231 or @abi or @c2c or @compile.sourcemap"
+	mvn test -Dtest=com.algorand.algosdk.integration.RunCucumberIntegrationTest -Dcucumber.filter.tags=$(INTEGRATION_TAGS)
+
+display-all-java-steps:
+	find src/test/java/com/algorand/algosdk -name *.java | xargs grep "io.cucumber.java.en" 2>/dev/null | grep -v Binary | cut -d: -f1 | sort | uniq | xargs grep -E "@(Given|Then|When)"
```
Comment (tzaffi) [inline Makefile:11]: Handy command I'm arguing to keep around that generates a file that I use as part of a fuzzy-logic step matching script.

Comment (tzaffi) [reply Makefile:11]: (But it's not as accurate as a pure java version would be)

File: Makefile [hunk 11]
```diff
@@ -1,10 +1,24 @@
...
+	mvn test -Dcucumber.filter.tags=$(UNIT_TAGS)
 
 integration:
-	mvn test \
-          -Dtest=com.algorand.algosdk.integration.RunCucumberIntegrationTest \
-          -Dcucumber.filter.tags="@algod or @assets or @auction or @kmd or @send or @send.keyregtxn or @indexer or @rekey_v1 or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @indexer.231 or @abi or @c2c or @compile.sourcemap"
+	mvn test -Dtest=com.algorand.algosdk.integration.RunCucumberIntegrationTest -Dcucumber.filter.tags=$(INTEGRATION_TAGS)
+
+display-all-java-steps:
+	find src/test/java/com/algorand/algosdk -name "*.java" | xargs grep "io.cucumber.java.en" 2>/dev/null | grep -v Binary | cut -d: -f1 | sort | uniq | xargs grep -E "@(Given|Then|When)"
```
Comment (tzaffi) [inline Makefile:11]: Command that generates input for **Unused Steps Analsysis** script

**Pull request merged**